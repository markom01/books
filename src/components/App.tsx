import React, { createContext, useEffect, useState } from "react";
import Home from "../routes/Home";

import { createBrowserRouter, RouterProvider } from "react-router-dom";

import { TableCellProps } from "@mui/material/TableCell";
import EditPage from "../routes/EditPage";
import ErrorPage from "../routes/errorPage";
import "../styles/App.scss";

interface HeaderDataTypes {
  align: TableCellProps["align"];
  label: "Title" | "Author" | "Year" | "Pages" | "Quantity";
}

export interface BookTypes {
  id: number;
  isbn: number;
  title: string;
  nameOfAuthor: string;
  dateOfBirthAuthor: Date;
  numOfPages: number;
  yearOfBublishing: number;
  quantity: number;
  coverPhoto: string;
}

export interface DataTypes {
  records: BookTypes[];
  start: number;
  limit: number;
  totalRecords: number;
  error: string;
  isLoading: boolean;
  selectedAuthor: string;
  isPreviewVisible: boolean;
  selectedBooks: BookTypes[];
  mutatedBooksCount: number;
  bookPreviewData: BookTypes;
  headerData: HeaderDataTypes[];
}

type ContextType = [DataTypes, React.Dispatch<React.SetStateAction<DataTypes>>];

const initBooks: DataTypes = {
  records: [],
  start: 0,
  limit: 0,
  totalRecords: 0,
  error: "",
  isLoading: true,
  selectedAuthor: "Any Author",
  selectedBooks: [],
  mutatedBooksCount: 0, //edited, deleted, added
  isPreviewVisible: false,
  bookPreviewData: {
    id: 0,
    isbn: 0,
    title: "",
    nameOfAuthor: "",
    dateOfBirthAuthor: new Date(),
    numOfPages: 0,
    yearOfBublishing: 0,
    quantity: 0,
    coverPhoto: "",
  },
  headerData: [
    { align: "left", label: "Title" },
    { align: "left", label: "Author" },
    { align: "left", label: "Year" },
    { align: "center", label: "Pages" },
    { align: "center", label: "Quantity" },
  ],
};

export const Context = createContext<ContextType>({} as ContextType);

const router = createBrowserRouter([
  {
    path: "/",
    element: <Home />,
    errorElement: <ErrorPage />,
  },
  {
    path: "/edit",
    element: <EditPage />,
  },
]);

function App() {
  const [books, setBooks] = useState<DataTypes>(initBooks);

  useEffect(() => {
    fetch(`https://book-store.mvsoft.co.rs/books`, {
      method: "GET",
      headers: {
        Accept: "application/json",
      },
    })
      .then((r) => r.json())
      .then((d) => {
        setBooks((prev) => ({
          ...prev,
          ...d,
          selectedBooks: d.records,
        }));
      })
      .catch((error) => {
        setBooks((prev) => ({
          ...prev,
          error: error.message,
        }));
      })
      .finally(() => {
        setBooks((prev) => ({
          ...prev,
          isLoading: false,
        }));
      });
  }, [books.mutatedBooksCount]);

  return (
    <Context.Provider value={[books, setBooks]}>
      <RouterProvider router={router} />
    </Context.Provider>
  );
}

export default App;
