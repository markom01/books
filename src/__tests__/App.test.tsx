import ReactDOM from "react-dom/client";
import renderer from "react-test-renderer";
import App from "../components/App";
import TestComponent from "../components/TestComponent";

describe("Unit test", () => {
  it("App coomponent renders without crashing", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    root.render(<App />);
  });
});

describe("Snapshot test", () => {
  it("Test component renders correctly", () => {
    const tree = renderer.create(<TestComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
