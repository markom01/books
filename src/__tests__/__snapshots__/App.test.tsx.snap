// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Test component renders correctly 1`] = `
<div>
  Test
</div>
`;
